### PR TITLE
change: [M3-7917] - TS Performance: Interfaces Over Intersections

### PR DESCRIPTION
**Pull request by abailly-akamai**
## Description 📝
This PR is a Typescript only PR to replace Intersections with Interfaces w/ `extends`. A Friday project that was a little more than I was aiming for 😄. 

- It should not have introduce any functionality deprecation
- Shouldn't weaken type safety
- Should improve TS performance (perhaps so very slightly, most notably in the IDE). Nice cleanup anyway.

📖 Resources on this topic:
- https://www.totaltypescript.com/react-apps-ts-performance
- https://github.com/microsoft/TypeScript/wiki/Performance#preferring-interfaces-over-intersections

⚠️ **The PR does not intent to fix poorly typed components and utils, only focusing on replacing intersections**

## Changes  🔄
- replace Intersections with Interfaces
- cleanup types where applicable and easy
- improve some naming conventions
- Updated docs (shamelessly stolen from the [Typescript wiki](https://github.com/microsoft/TypeScript/wiki/Performance#preferring-interfaces-over-intersections))

## How to test 🧪

- Pull code locally
- Run `yarn tsc`
- Run `yarn test`

## As an Author I have considered 🤔

*Check all that apply*

- [ ] 👀 Doing a self review
- [ ] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [ ] 🤏 Splitting feature into small PRs
- [x] ➕ Adding a changeset
- [ ] 🧪 Providing/Improving test coverage
- [ ] 🔐 Removing all sensitive information from the code and PR description
- [ ] 🚩 Using a feature flag to protect the release
- [ ] 👣 Providing comprehensive reproduction steps
- [ ] 📑 Providing or updating our documentation
- [ ] 🕛 Scheduling a pair reviewing session
- [ ] 📱 Providing mobile support
- [ ] ♿  Providing accessibility support

### CODE REVIEW

File: packages/manager/src/containers/account.container.ts
Comment (bnussman-akamai) [inline packages/manager/src/containers/account.container.ts:1]: What's the motivation behind the changes to the generics?

Comment (abailly-akamai) [reply packages/manager/src/containers/account.container.ts:1]: same motivation as to removing non-needed intersections. Do you see any drawback to this?

**Comment by carrillo-erik** [comment]
@abailly-akamai I'm getting an error when trying to run this locally. Is there any other commands or flags we need to pass to run this locally?

1. I checked out this PR
2. ran `yarn tsc`
3. Screenshot 2024-03-25 at 10 47 12 AM

File: packages/manager/src/containers/account.container.ts
Comment (bnussman-akamai) [reply packages/manager/src/containers/account.container.ts:1]: I'm mainly wondering if `React.ComponentType<Props & WithAccountProps>` ➡️ `React.ComponentType<WithAccountProps>` changes anything significant

Comment (abailly-akamai) [reply packages/manager/src/containers/account.container.ts:1]: it looks to achieve parity and those are frankly not really typed that strongly anyway - open to suggestions or reverting if there's concerns, but from my local testing + e2e we seem to look good

maybe @jaalah-akamai can double check since he's a stronger typescripter than I am

**Comment by abailly-akamai** [comment]
@carrillo-erik run `yarn tsc` from packages/manager

**Review by bnussman-akamai — APPROVED** [review]

**Comment by carrillo-erik** [comment]
> @carrillo-erik run `yarn tsc` from packages/manager

@abailly-akamai Thanks, that resolved the error with running `yarn tsc`. Now I'm seeing two errors with the tests after running `yarn test`. Do you get the same errors locally or are you aware of this?

Screenshot 2024-03-25 at 12 00 00 PM

**Comment by abailly-akamai** [comment]
@carrillo-erik am not getting those errors, nor is CI. See if you are getting them on the `develop` branch as well. Maybe something with your setup

**Review by dwiley-akamai — APPROVED** [review]
`yarn tsc` ✅
`yarn test` ✅ (I did see some "Unhandled Errors"/"Unhandled Rejections" locally, but all the tests passed remotely)
Code review ✅
Didn't observe any issues in app ✅

**Review by carrillo-erik — APPROVED** [review]

**Comment by carrillo-erik** [comment]
> @carrillo-erik am not getting those errors, nor is CI. See if you are getting them on the `develop` branch as well. Maybe something with your setup

I've ran the app and inspected the IDE performance without further issue. I'm comfortable with all the test passing on this PR with our pipelines. After pulling down the latest from develop and running the test, I experienced the same failures, meaning the failures are not a cause of these changes. I've approved your PR ✅✅✅

**Review by jaalah-akamai — APPROVED** [review]
Overall nice work and cleanup - just left one comment for consideration 👍

File: packages/manager/src/containers/account.container.ts
Comment (abailly-akamai) [reply packages/manager/src/containers/account.container.ts:1]: All fair. I reverted all HOC changes - those will eventually go gradually anyway

File: docs/development-guide/13-coding-standards.md [hunk 1]
```diff
@@ -132,6 +132,31 @@ myFunction(CreateTypes.Backup); // Works
...
+For this reason, extending types with interfaces/extends is suggested over creating intersection types.
+
+```Typescript
+- type Foo = Bar & Baz & {
+-     someProp: string;
+- }
++ interface Foo extends Bar, Baz {
++     someProp: string;
++ }
+```
```
[the remark this answers is not in the document]
Comment (abailly-akamai) [reply docs/development-guide/13-coding-standards.md:158]: Good point - done!
